PersistentMockRegistry: Prune deactivated PersistentMocks after reaching a soft limit

copperwall opened the pull request:
This adds functionality to remove deactivated persistent mocks
from the PersistentMockRegistry once the number of registered
mocks exceeds REGISTRY_SOFT_LIMIT (in this case 100).

copperwall committed:
feat: prune deactivated PersistentMocks

This adds functionality to remove deactivated persistent mocks
from the PersistentMockRegistry once the number of registered
mocks exceeds REGISTRY_SOFT_LIMIT (in this case 100).

lexidor commented:
Thank you for your PR and participation.

I have a suggestion to improve performance somewhat.
As it is right now, you'll iterate over the vec and allocate a new one on every call to `register` where the stored count exceeds the constant.
This could be very wasteful if new mocks were being added without some of them being disabled.
If you have `n` mocks and need to add `m` more, the time complexity of this would be `O(n*m)`.
It might be a plan to increment a class variable each time a mock is added and only run cleanup on every n'th call.
This would make the cost of adding a new mock less expensive, trading of some RAM for a lot less CPU time.

tyronjung-quizlet commented on tests/PersistentTest.php:
```diff
@@ -378,3 +394,19 @@ private function noopGlobalFunction(): PersistentFunctionMock {
+	}
+}
+
+class MockDeactivatable implements Interfaces\IDeactivatable {
```
Would you be able to group these test classes under `tests/Fixtures/` along with the rest of the test classes?

copperwall replied:
Yeah definitely!

tyronjung-quizlet reviewed (commented):
Great to see unit tests. 👍

copperwall commented:
@lexidor Thanks for the feedback! That sounds like a great idea, I'll give that a shot.

copperwall committed:
fix: add MockPersistentMockRegistry to tests/Fixtures

copperwall commented:
I'm still getting around to the performance improvements, but for now I thought I'd push up what I have for moving the fixture classes to the Fixtures directory.

copperwall committed:
feat: only remove mocks every PRUNE_INTERVAL attempts past REGISTRY_SOFT_LIMIT

lexidor commented on src/Persistent/PersistentMockRegistry.php:
```diff
@@ -2,17 +2,31 @@
+		//
+		// If the number of items in the registry drops below REGISTRY_SOFT_LIMIT,
+		// the number of attempts is reset to zero.
+		if (C\count(self::$registry) >= self::REGISTRY_SOFT_LIMIT &&
```
I think you have made a coding error here.
The first pass you prune and `++` the pruneAttempts.
The next pass, `self::$pruneAttempts % self::PRUNE_INTERVAL === 0` is false, so you fall into the else case.
The else resets pruneAtetmpts to 0.
Wouldn't this create a cycle like this `0 -> 1 -> 0 -> 1 -> 0` for pruneAttempts?

copperwall replied:
Oh geez, thanks for pointing that out @lexidor. I didn't quite think that through. I just pushed up a fix so that it attempts every PRUNE_INTERVAL times while the registry is over the soft limit. I picked 10 for now as a constant, but I'm open to talking about that

copperwall committed:
fix: only reset pruneAttempts when under soft limit

This changes the if condition for pruning in PersistentMockRegistry::register
to update the pruneAttempts counter when the registry size is greater than or equal to the
soft limit, and not just when the size is greater than the limit *and* the pruneAttempts
counter is divisible by PRUNE_INTERVAL.

This also changes the pruneAttempts reset behavior to reset whenever the registry
size is less than the soft limit.

lexidor commented on src/Persistent/PersistentMockRegistry.php:
```diff
@@ -2,17 +2,33 @@
 use type Hammock\Interfaces\IDeactivatable;
 
 class PersistentMockRegistry {
+	const REGISTRY_SOFT_LIMIT = 100;
```
Does this typecheck in `// strict` mode?
Shouldn't it require it to be `const int REGISTRY_SOFT_LIMIT = 100`?

copperwall replied:
That's a good question.

I found some docs in the Hack Constants Docs that say
> If a class constant's type is omitted, it is inferred from the initializer, which must be present. In this case, that type is string.

https://docs.hhvm.com/hack/classes/constants

So I guess since the initializer is an int constant it infers that.

Explicitly adding types sounds nice though, I pushed a commit with some int types.

lexidor reviewed (approved):
The logic is now correct.
Sorry for falling of the face of the earth.
I just have one quick question about the typechecker.

copperwall committed:
fix: add int types to class constants

copperwall commented:
No worries! I pushed up a commit that adds some explicit int types.

copperwall commented:
Sorry for the force push, it looked liked one of the build jobs didn't run correctly so I was trying to retrigger the travis job

tyronjung-quizlet commented:
Looks good! Not sure what's going on with Travis, but I'm going to go ahead and merge this change.
